src/core/lib/channel/channel_tracer.c

```diff
@@ -0,0 +1,325 @@
+#endif
+  grpc_channel_tracer* tracer = gpr_zalloc(sizeof(grpc_channel_tracer));
+  gpr_mu_init(&tracer->tracer_mu);
+  gpr_ref_init(&tracer->refs, 1);
+#ifdef GRPC_CHANNEL_TRACER_REFCOUNT_DEBUG
+  gpr_log(GPR_DEBUG, "%p create [%s:%d %s]", tracer, file, line, func);
+#endif
+  tracer->node_list.max_size = max_nodes;
+  tracer->time_created = gpr_now(GPR_CLOCK_REALTIME);
+  return tracer;
+}
+
+#ifdef GRPC_CHANNEL_TRACER_REFCOUNT_DEBUG
+grpc_channel_tracer* grpc_channel_tracer_ref(grpc_channel_tracer* tracer,
+                                             const char* file, int line,
+                                             const char* func) {
+  if (!tracer) return tracer;
+  gpr_log(GPR_DEBUG, "%p: %" PRIdPTR " -> %" PRIdPTR " [%s:%d %s]", tracer,
+          gpr_atm_no_barrier_load(&tracer->refs.count),
+          gpr_atm_no_barrier_load(&tracer->refs.count) + 1, file, line, func);
+  gpr_ref(&tracer->refs);
+  return tracer;
+}
+#else
+grpc_channel_tracer* grpc_channel_tracer_ref(grpc_channel_tracer* tracer) {
+  if (!tracer) return tracer;
+  gpr_ref(&tracer->refs);
+  return tracer;
+}
+#endif
+
+static void free_node(grpc_trace_node* node) {
+  // no need to free string, since they are always static
+  GRPC_ERROR_UNREF(node->error);
+  if (node->subchannel) {
+    GRPC_CHANNEL_TRACER_UNREF(node->subchannel);
+  }
+  gpr_free(node);
+}
+
+void grpc_channel_tracer_destroy(grpc_channel_tracer* tracer) {
+  grpc_trace_node* it = tracer->node_list.head_trace;
+  while (it) {
+    grpc_trace_node* to_free = it;
+    it = it->next;
+    free_node(to_free);
+  }
+  gpr_free(tracer);
+}
+
+#ifdef GRPC_CHANNEL_TRACER_REFCOUNT_DEBUG
+void grpc_channel_tracer_unref(grpc_channel_tracer* tracer, const char* file,
+                               int line, const char* func) {
+  if (!tracer) return;
+  gpr_log(GPR_DEBUG, "%p: %" PRIdPTR " -> %" PRIdPTR " [%s:%d %s]", tracer,
+          gpr_atm_no_barrier_load(&tracer->refs.count),
+          gpr_atm_no_barrier_load(&tracer->refs.count) - 1, file, line, func);
+  if (gpr_unref(&tracer->refs)) {
+    grpc_channel_tracer_destroy(tracer);
+  }
+}
+#else
+void grpc_channel_tracer_unref(grpc_channel_tracer* tracer) {
+  if (!tracer) return;
+  if (gpr_unref(&tracer->refs)) {
+    grpc_channel_tracer_destroy(tracer);
+  }
+}
+#endif
+
+static void add_trace(grpc_trace_node_list* list, const char* trace,
+                      grpc_error* error,
+                      grpc_connectivity_state connectivity_state,
+                      grpc_channel_tracer* subchannel) {
+  grpc_trace_node* new_trace_node = gpr_malloc(sizeof(grpc_trace_node));
+  new_trace_node->data = trace;
+  new_trace_node->error = error;
+  new_trace_node->time_created = gpr_now(GPR_CLOCK_REALTIME);
+  new_trace_node->connectivity_state = connectivity_state;
+  new_trace_node->next = NULL;
+  new_trace_node->subchannel = subchannel;
+
+  if (subchannel) {
+    GRPC_CHANNEL_TRACER_REF(subchannel);
```

Comment:
Make a decision here: is it safe to ref a null node (remove this if) or it's unsafe (remove the one in ref)